rpcs3/ds3_pad_handler.cpp

```diff
@@ -0,0 +1,433 @@
+	pad->m_sticks.emplace_back(CELL_PAD_BTN_OFFSET_ANALOG_RIGHT_X, FindKeyCode(button_list, p_profile->rs_left), FindKeyCode(button_list, p_profile->rs_right));
+	pad->m_sticks.emplace_back(CELL_PAD_BTN_OFFSET_ANALOG_RIGHT_Y, FindKeyCode(button_list, p_profile->rs_down), FindKeyCode(button_list, p_profile->rs_up));
+
+	pad->m_vibrateMotors.emplace_back(true, 0);
+	pad->m_vibrateMotors.emplace_back(false, 0);
+
+	bindings.emplace_back(ds3device, pad);
+
+	return true;
+}
+
+void ds3_pad_handler::ThreadProc()
+{
+	for (int i = 0; i < static_cast<int>(bindings.size()); i++)
+	{
+		m_dev = bindings[i].first;
+		auto thepad = bindings[i].second;
+		auto profile = m_dev->config;
+
+		thepad->m_port_status = CELL_PAD_STATUS_CONNECTED | CELL_PAD_STATUS_ASSIGN_CHANGES;
+
+		if (get_data(m_dev))
+		{
+			if (get_data(m_dev))
+				process_data(m_dev, thepad);
+		}
+		if (m_dev->large_motor != thepad->m_vibrateMotors[0].m_value || m_dev->small_motor != thepad->m_vibrateMotors[1].m_value)
+		{
+			m_dev->large_motor = thepad->m_vibrateMotors[0].m_value;
+			m_dev->small_motor = thepad->m_vibrateMotors[1].m_value;
+			send_output_report(m_dev);
+		}
+	}
+}
+
+void ds3_pad_handler::TestVibration(const std::string& padId, u32 largeMotor, u32 smallMotor)
+{
+	std::shared_ptr<ds3_device> device = get_device(padId);
+	if (device == nullptr || device->handle == nullptr)
+		return;
+
+	// Set the device's motor speeds to our requested values 0-255
+	device->large_motor = largeMotor;
+	device->small_motor = smallMotor;
+
+	int index = 0;
+	for (int i = 0; i < MAX_GAMEPADS; i++)
+	{
+		if (g_cfg_input.player[i]->handler == pad_handler::ds3)
+		{
+			if (g_cfg_input.player[i]->device.to_string() == padId)
+			{
+				m_pad_configs[index].load();
+				device->config = &m_pad_configs[index];
+				break;
+			}
+			index++;
+		}
+	}
+
+	// Start/Stop the engines :)
+	send_output_report(device);
+}
+
+void ds3_pad_handler::GetNextButtonPress(const std::string& padId, const std::function<void(u16, std::string, std::string, int[])>& callback, const std::function<void(std::string)>& fail_callback, bool get_blacklist, const std::vector<std::string>& buttons)
+{
+	if (get_blacklist)
+		blacklist.clear();
+
+	std::shared_ptr<ds3_device> device = get_device(padId);
+	if (device == nullptr || device->handle == nullptr)
+		return fail_callback(padId);
+
+	return;
+}
+
+void ds3_pad_handler::send_output_report(const std::shared_ptr<ds3_device>& ds3dev)
+{
+	u8 report_buf[] = {
+		0x00, 0xff, 0x00, 0xff, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0x00, 0x00, 0x00, 0x00, 0x00
+	};
+
+	report_buf[2] = ds3dev->large_motor;
+	report_buf[4] = ds3dev->small_motor;
+
+	libusb_control_transfer(ds3dev->handle, LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE, 0x09, 0x0201, 0, report_buf, sizeof(report_buf), 0);
+}
```

Comment:
consider adding explicit enums or even a single macro to help with readability. I'm talking about the raw hex values 0x09 and 0x0201